scripts/utils.py

```diff
@@ -2,19 +2,176 @@
 Shared functions between various scripts
 """
 import re
+import sys
+
+try:
+    import tomllib
+except ModuleNotFoundError:
+    # pre-Python 3.11 compatibility
+    import toml as tomllib
 from pathlib import Path
 
 RE_NAME = re.compile(r'name="(.*?)"')
 
 
-def project_name(path: Path) -> str:
+def get_project_name(path: Path) -> str:
     """Extract the project name from setup.py"""
     setup_py = path / "setup.py"
     if not setup_py.exists():
-        raise RuntimeError(f"No setup.py in {path}."
-                           "If this isn't a Python project, use --project?")
+        raise RuntimeError(
+            f"No setup.py in {path}." "If this isn't a Python project, use --project?"
+        )
     search = RE_NAME.search(setup_py.read_text())
     if not search:
-        raise RuntimeError(f"Unable to parse name out of {setup_py}. "
-                           "If this isn't a Python project, use --project?")
+        raise RuntimeError(
+            f"Unable to parse name out of {setup_py}. "
+            "If this isn't a Python project, use --project?"
+        )
     return search.group(1)
+
+
+def get_requirements_names_and_versions(path: Path) -> list[(str, str)]:
+    """
+    Return a list of package names and versions for all dependencies declared in a
+    requirements.txt file.
+    """
+    ret = []
+    for line in path.read_text().splitlines():
+        if line.startswith("#"):
+            continue
+        if "==" not in line:
+            continue
+        name, constraint = line.split("==", 1)
+        version = constraint.rstrip("\\").strip()
+        ret.append((name, version))
+    return ret
+
+
+def get_poetry_names_and_versions(
+    path_to_poetry_lock: Path, path_to_pyproject_toml
+) -> list[(str, str)]:
+    """
+    Return a list of package names and versions for all main (non-development) dependencies,
+    including transitive ones, defined via Poetry configuration files.
+    """
+    data = tomllib.loads(path_to_poetry_lock.read_text())
+    relevant_dependencies = get_relevant_poetry_dependencies(
+        path_to_pyproject_toml, path_to_poetry_lock
+    )
+    ret = []
+    for package in data["package"]:
+        if package["name"] not in relevant_dependencies:
+            continue
+        ret.append((package["name"], package["version"]))
+
+    return ret
+
+
+def get_relevant_poetry_dependencies(
+    path_to_pyproject_toml: Path, path_to_poetry_lock: Path
+) -> list[str]:
+    """
+    Identify main (non-development) requirements. poetry.lock does not preserve
+    the distinction between different dependency groups, so we have to parse
+    both files.
+    """
+    pyproject_dict = tomllib.loads(path_to_pyproject_toml.read_text())
+
+    # Extract main dependencies
+    main_dependencies = list(pyproject_dict["tool"]["poetry"]["dependencies"].keys())
```

Comment:
nit: You don't need `.keys()`, running `list()` over a dict will only return keys.

Reply:
We're modifying the keys as well; I've changed the code to do that on the dict instead.